apps/system/js/updatable.js

```diff
@@ -70,16 +91,103 @@ AppUpdatable.prototype.clean = function() {
 AppUpdatable.prototype.availableCallBack = function() {
   this.size = this.app.downloadSize;
 
-  if (this.app.installState === 'installed') {
-    UpdateManager.addToUpdatesQueue(this);
-
-    // we add these callbacks only now to prevent interfering
-    // with other modules (especially the AppInstallManager)
-    this.app.ondownloaderror = this.errorCallBack.bind(this);
-    this.app.ondownloadsuccess = this.successCallBack.bind(this);
-    this.app.ondownloadapplied = this.appliedCallBack.bind(this);
-    this.app.onprogress = this.progressCallBack.bind(this);
+  if (this.app.installState !== 'installed') {
+    return;
   }
+
+  let settings = [
+    setting(AppUpdatable.AUTO_UPDATE),
+    setting(AppUpdatable.NOTIFY_UPDATE)
+  ];
+
+  Promise.all(settings).then(results => {
+
+    var autoAddonUpdates = results[0];
+    var notifyAddonUpdates = results[1];
+
+    // If we have autoUpdates turned off for addons, they
+    // go via the usual install flow
+    if (this.app.manifest.role === 'addon' && autoAddonUpdates) {
+
+      // We fail silently, the addon will be updated next time
+      // we check for updates
+      this.app.ondownloaderror = err => {
+        console.error('zxc failed to auto update', this.name,
+                      err.application.downloadError.name);
+      };
+
+      this.app.ondownloadsuccess = this.applyUpdate.bind(this);
+      this.app.onprogress = null;
+      this.app.ondownloadapplied = null;
+
+      if (notifyAddonUpdates) {
+        this.app.ondownloadapplied = this.notifyAutoUpdate.bind(this);
+      }
+
+      this.app.download();
+      this.maybeNotifyUpdate();
+    } else {
+      UpdateManager.addToUpdatesQueue(this);
+
+      // we add these callbacks only now to prevent interfering
+      // with other modules (especially the AppInstallManager)
+      this.app.ondownloaderror = this.errorCallBack.bind(this);
+      this.app.ondownloadsuccess = this.successCallBack.bind(this);
+      this.app.ondownloadapplied = this.appliedCallBack.bind(this);
+      this.app.onprogress = this.progressCallBack.bind(this);
+    }
+  });
+};
+
+AppUpdatable.prototype.notifyAutoUpdate = function() {
+  IconsHelper.getIcon(this.app.origin, 32, null, this.app).then(iconUrl => {
+    var opts = {
+      tag: 'addonUpdated',
```

Comment:
This is a little weird. Do we want separate notifications for each add-on? Right now only the last applied add-on will have the notifications since it will overwrite earlier add-on notifications. Not sure if this is the behavior we want here.